### ISSUE TITLE
#unpublishgate

### ISSUE THREAD

**Issue by bsiddiqui** [open-sorcery/discuss#8]
http://www.haneycodes.net/npm-left-pad-have-we-forgotten-how-to-program/

Thoughts?

**Comment by bendrucker** [comment]
Can't really find anything to agree with there. Writing for loops is a skill, just not one I care about having. 

Couple comments put it well:

First:

> Yeah, I think you’re just fundamentally misunderstanding the npm ecosystem and small module philosophy. +1 on the above link from Luis. The real problem yesterday’s events exposed is a design flaw in npm. Immutability at the centralized authority level and more decentralization of package distribution is the solution, not “write more functions yourself.”

Second:

> These packages proliferate because JavaScript’s standard library is so woefully inadequate. It’s ridiculous that anyone has to write their own left-pad function in the first place. The problem isn’t that people “don’t know how to code”, it’s that they’re operating in an environment where basic functionality has to be implemented by end-users (i.e. developers), and where it’s — perhaps reasonably — standard operating procedure to import someone else’s solution rather than write your own.

Author seems like he's written some good C#. That's a very different world. He hasn't written any open source JS.

**Comment by chinedufn** [comment]
Like Ben said, when reading this it didn't feel like he'd written much OS JS.

This inspires me to use more things outside of my normal toolkit. I'm probably just as ignorant in my thinking about things that I've never done.

Maybe I should spend an evening here and there writing some `C#`, etc.

How about your thoughts @bsiddiqui ?

**Comment by bsiddiqui** [comment]
I don't agree with the article's suggestions: write all inline functions, don't rely on small modules that are less than "x" LOC, etc. I do think, however, that having an ecosystem where lots of packages are reliant on obscure, likely untested code, with no guarantee the dependencies won't change is a bit frightening. 

Issues with packages being taken down, versioned incorrectly, etc creates points of failure that anyone doing professional software dev should be weary of. Recently I had an issue where Lob made a breaking change to their package and did a minor release, leading to some unexpected behavior.

I agree with the perspective that it's especially concerning if you expose these failure points for 1 LOC. 

I'll likely just be much more careful about what software I depend on - something that everyone should be doing regardless. The npm ecosystem makes it simple to just find a package that does "x", require it, and ask questions later. 

So basically totally fine to not want to write for utility functions and depend on abstractions as long as you know exactly what it's doing, it's well tested, and you have some degree of confidence that the maintainer won't take it down on a whim. 

-- 

What would you guys think about NPM requiring some amount of notice be given before taking down a package? Perhaps if a package has dependencies you a notification could go out to maintainers on dependent projects that the package will be taken down in "x" days. 

It'd leave devs with the autonomy to do whatever with their software and would give others some assurance that a package won't randomly disappear. Can't imagine anyone writing OSS having an issue with that.

**Comment by bendrucker** [comment]
> What would you guys think about NPM requiring some amount of notice be given before taking down a package?

Nice in theory, doubt it could be executed well. I think the move here is full immutability. Just don't let people unpublish public packages without contacting support.
